### Here you can also see a super simple test:

```python
import asyncio
from pydantic import BaseModel
from any_llm import AnyLLM, LLMProvider


class City(BaseModel):
    city_name: str


async def main() -> None:
    llm = AnyLLM.create(LLMProvider.OLLAMA)
    chunks = await llm.acompletion(
        model="llama3.2:1b",
        messages=[{"role": "user", "content": "What is the capital of France?"}],
        response_format=City,
        stream=True,
    )
    content = "".join([c.choices[0].delta.content or "" async for c in chunks])
    print(f"Raw streamed output: {content!r}")
    print(f"Parsed: {City.model_validate_json(content)}")


asyncio.run(main())
```

#### Executing using my feature branch:

```bash
❯ uv run --with "any-llm-sdk[ollama] @ git+https://github.com/jonigl/any-llm.git@feat/ollama-format" python test.py
    Updated https://github.com/jonigl/any-llm.git (c28685a6641937bf93a688f0165a436876f83817)
      Built any-llm-sdk @ git+https://github.com/jonigl/any-llm.git@c28685a6641937bf93a688f0165a436876f83817
Installed 25 packages in 112ms
Raw streamed output: '{\n    "city_name": "Paris"\n}'
Parsed: city_name='Paris'
```

#### Executing using any-llm `v1.19.0`

```bash
❯ uv run --with "any-llm-sdk[ollama] @ git+https://github.com/mozilla-ai/any-llm.git@1.19.0" python test.py
    Updated https://github.com/mozilla-ai/any-llm.git (ca0ef070210982af672111397f0efce781dae073)
      Built any-llm-sdk @ git+https://github.com/mozilla-ai/any-llm.git@ca0ef070210982af672111397f0efce781dae073
Installed 25 packages in 167ms
Raw streamed output: 'The capital of France is Paris.'
Traceback (most recent call last):
  File "/Users/jlowenstern/jonigl/ai/repos/any-llm/test.py", line 23, in <module>
    asyncio.run(main())
  File "/Users/jlowenstern/.local/share/uv/python/cpython-3.12.7-macos-aarch64-none/lib/python3.12/asyncio/runners.py", line 194, in run
    return runner.run(main)
           ^^^^^^^^^^^^^^^^
  File "/Users/jlowenstern/.local/share/uv/python/cpython-3.12.7-macos-aarch64-none/lib/python3.12/asyncio/runners.py", line 118, in run
    return self._loop.run_until_complete(task)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jlowenstern/.local/share/uv/python/cpython-3.12.7-macos-aarch64-none/lib/python3.12/asyncio/base_events.py", line 687, in run_until_complete
    return future.result()
           ^^^^^^^^^^^^^^^
  File "/Users/jlowenstern/jonigl/ai/repos/any-llm/test.py", line 20, in main
    print(f"Parsed: {City.model_validate_json(content)}")
                     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/jlowenstern/.cache/uv/archive-v0/pRcIa_nqO-BXYhAMvIE16/lib/python3.12/site-packages/pydantic/main.py", line 782, in model_validate_json
    return cls.__pydantic_validator__.validate_json(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
pydantic_core._pydantic_core.ValidationError: 1 validation error for City
  Invalid JSON: expected value at line 1 column 1 [type=json_invalid, input_value='The capital of France is Paris.', input_type=str]
    For further information visit https://errors.pydantic.dev/2.13/v/json_invalid
```

--- a/src/any_llm/providers/ollama/ollama.py
+++ b/src/any_llm/providers/ollama/ollama.py
@@ -137,7 +137,11 @@ class OllamaProvider(AnyLLM):
         if isinstance(response_format, dict):
             response_type = response_format.get("type")
             if response_type == "json_schema":
-                schema: dict[str, Any] = response_format["json_schema"]["schema"]
+                json_schema = response_format.get("json_schema")
+                if json_schema is None or "schema" not in json_schema:
+                    msg = "json_schema response_format must include 'json_schema.schema'"
+                    raise ValueError(msg)
+                schema: dict[str, Any] = json_schema["schema"]
                 return schema
             if response_type == "json_object":
                 return "json"

--- a/src/any_llm/providers/ollama/ollama.py
+++ b/src/any_llm/providers/ollama/ollama.py
@@ -1,7 +1,7 @@
 from __future__ import annotations
 
 import json
-from typing import TYPE_CHECKING, Any
+from typing import TYPE_CHECKING, Any, Literal
 
 from typing_extensions import override
 
@@ -119,6 +119,33 @@ class OllamaProvider(AnyLLM):
         """Convert Ollama list models response to OpenAI format."""
         return _convert_models_list(response)
 
+    @staticmethod
+    def _convert_response_format(
+        response_format: dict[str, Any] | type | None,
+    ) -> Literal["json"] | dict[str, Any] | None:
+        """Convert a `response_format` into Ollama's `format` argument.
+
+        Ollama's `format` accepts either the string "json" (unconstrained JSON mode) or a raw
+        JSON schema dict (https://docs.ollama.com/capabilities/structured-outputs). OpenAI-style
+        dicts (`json_object`, `json_schema`, `text`) are translated accordingly; any other dict
+        is assumed to already be a raw schema and passed through unchanged.
+        """
+        if response_format is None:
+            return None
+        if is_structured_output_type(response_format):
+            return get_json_schema(response_format)
+        if isinstance(response_format, dict):
+            response_type = response_format.get("type")
+            if response_type == "json_schema":
+                schema: dict[str, Any] = response_format["json_schema"]["schema"]
+                return schema
+            if response_type == "json_object":
+                return "json"
+            if response_type == "text":
+                return None
+            return response_format
+        return None
+
     @override
     def _init_client(self, api_key: str | None = None, api_base: str | None = None, **kwargs: Any) -> None:
         self.client = AsyncClient(host=api_base, **kwargs)
@@ -160,7 +187,7 @@ class OllamaProvider(AnyLLM):
         self,
         model: str,
         messages: list[dict[str, Any]],
-        output_format: dict[str, Any] | None = None,
+        output_format: Literal["json"] | dict[str, Any] | None = None,
         **kwargs: Any,
     ) -> AsyncIterator[ChatCompletionChunk]:
         """Handle streaming completion - extracted to avoid generator issues."""
@@ -199,12 +226,7 @@ class OllamaProvider(AnyLLM):
     ) -> ChatCompletion | AsyncIterator[ChatCompletionChunk]:
         """Create a chat completion using Ollama."""
 
-        output_format: dict[str, Any] | None = None
-        if params.response_format is not None:
-            if is_structured_output_type(params.response_format):
-                output_format = get_json_schema(params.response_format)
-            elif isinstance(params.response_format, dict):
-                output_format = params.response_format
+        output_format = self._convert_response_format(params.response_format)
 
         # (https://www.reddit.com/r/ollama/comments/1ked8x2/feeding_tool_output_back_to_llm/)
         cleaned_messages = []

--- a/src/any_llm/providers/ollama/ollama.py
+++ b/src/any_llm/providers/ollama/ollama.py
@@ -160,6 +160,7 @@ class OllamaProvider(AnyLLM):
         self,
         model: str,
         messages: list[dict[str, Any]],
+        output_format: dict[str, Any] | None = None,
         **kwargs: Any,
     ) -> AsyncIterator[ChatCompletionChunk]:
         """Handle streaming completion - extracted to avoid generator issues."""
@@ -169,6 +170,7 @@ class OllamaProvider(AnyLLM):
             messages=messages,
             tools=kwargs.pop("tools", None),
             think=kwargs.pop("think", None),
+            format=output_format,
             stream=True,
             options=kwargs,
         )
@@ -233,7 +235,9 @@ class OllamaProvider(AnyLLM):
         completion_kwargs = self._convert_completion_params(params, **kwargs)
 
         if params.stream:
-            return self._stream_completion_async(params.model_id, cleaned_messages, **completion_kwargs)
+            return self._stream_completion_async(
+                params.model_id, cleaned_messages, output_format=output_format, **completion_kwargs
+            )
 
         response: OllamaChatResponse = await self.client.chat(
             model=params.model_id,

--- a/tests/unit/providers/test_ollama_provider.py
+++ b/tests/unit/providers/test_ollama_provider.py
@@ -1,4 +1,5 @@
 from collections.abc import AsyncIterator
+from typing import Any
 from unittest.mock import AsyncMock, Mock, patch
 
 import pytest
@@ -154,6 +155,19 @@ async def test_completion_with_raw_schema_response_format() -> None:
 
             call_kwargs = provider.client.chat.call_args[1]
             assert call_kwargs["format"] == response_format
+
+
+@pytest.mark.parametrize(
+    "response_format",
+    [
+        {"type": "json_schema"},
+        {"type": "json_schema", "json_schema": {"name": "TestOutput"}},
+    ],
+)
+def test_convert_response_format_malformed_json_schema_raises(response_format: dict[str, Any]) -> None:
+    """Test that a json_schema response_format missing the inner schema raises a descriptive error."""
+    with pytest.raises(ValueError, match=r"must include 'json_schema\.schema'"):
+        OllamaProvider._convert_response_format(response_format)
 
 
 @pytest.mark.asyncio

--- a/tests/unit/providers/test_ollama_provider.py
+++ b/tests/unit/providers/test_ollama_provider.py
@@ -68,9 +68,75 @@ async def test_completion_with_dataclass_response_format() -> None:
 
 
 @pytest.mark.asyncio
-async def test_completion_with_dict_response_format() -> None:
-    """Test that dict response_format is passed through unchanged."""
-    response_format = {"type": "json_object"}
+async def test_completion_with_json_object_response_format() -> None:
+    """Test that OpenAI-style {"type": "json_object"} is translated to Ollama's "json" mode."""
+    with patch.object(OllamaProvider, "_init_client"):
+        provider = OllamaProvider(api_key=None)
+        provider.client = Mock()
+        provider.client.chat = AsyncMock(return_value=Mock())
+
+        with patch.object(OllamaProvider, "_convert_completion_response", return_value=Mock()):
+            await provider._acompletion(
+                CompletionParams(
+                    model_id="llama3.1",
+                    messages=[{"role": "user", "content": "Hello"}],
+                    response_format={"type": "json_object"},
+                ),
+            )
+
+            call_kwargs = provider.client.chat.call_args[1]
+            assert call_kwargs["format"] == "json"
+
+
+@pytest.mark.asyncio
+async def test_completion_with_json_schema_response_format() -> None:
+    """Test that OpenAI-style {"type": "json_schema", ...} has its inner schema extracted."""
+    schema = {"type": "object", "properties": {"name": {"type": "string"}}, "required": ["name"]}
+    response_format = {"type": "json_schema", "json_schema": {"name": "TestOutput", "schema": schema}}
+
+    with patch.object(OllamaProvider, "_init_client"):
+        provider = OllamaProvider(api_key=None)
+        provider.client = Mock()
+        provider.client.chat = AsyncMock(return_value=Mock())
+
+        with patch.object(OllamaProvider, "_convert_completion_response", return_value=Mock()):
+            await provider._acompletion(
+                CompletionParams(
+                    model_id="llama3.1",
+                    messages=[{"role": "user", "content": "Hello"}],
+                    response_format=response_format,
+                ),
+            )
+
+            call_kwargs = provider.client.chat.call_args[1]
+            assert call_kwargs["format"] == schema
+
+
+@pytest.mark.asyncio
+async def test_completion_with_text_response_format() -> None:
+    """Test that OpenAI-style {"type": "text"} disables Ollama's format constraint."""
+    with patch.object(OllamaProvider, "_init_client"):
+        provider = OllamaProvider(api_key=None)
+        provider.client = Mock()
+        provider.client.chat = AsyncMock(return_value=Mock())
+
+        with patch.object(OllamaProvider, "_convert_completion_response", return_value=Mock()):
+            await provider._acompletion(
+                CompletionParams(
+                    model_id="llama3.1",
+                    messages=[{"role": "user", "content": "Hello"}],
+                    response_format={"type": "text"},
+                ),
+            )
+
+            call_kwargs = provider.client.chat.call_args[1]
+            assert call_kwargs["format"] is None
+
+
+@pytest.mark.asyncio
+async def test_completion_with_raw_schema_response_format() -> None:
+    """Test that a raw JSON schema dict (Ollama's native style) is passed through unchanged."""
+    response_format = {"type": "object", "properties": {"name": {"type": "string"}}, "required": ["name"]}
 
     with patch.object(OllamaProvider, "_init_client"):
         provider = OllamaProvider(api_key=None)
@@ -123,6 +189,34 @@ async def test_streaming_completion_passes_format_top_level() -> None:
         assert isinstance(call_kwargs["format"], dict)
         assert "name" in call_kwargs["format"]["properties"]
         assert "format" not in call_kwargs.get("options", {})
+
+
+@pytest.mark.asyncio
+async def test_streaming_completion_translates_json_object_format() -> None:
+    """Test that OpenAI-style {"type": "json_object"} is translated to "json" when streaming."""
+
+    async def empty_async_iter() -> AsyncIterator[None]:
+        return
+        yield
+
+    with patch.object(OllamaProvider, "_init_client"):
+        provider = OllamaProvider(api_key=None)
+        provider.client = Mock()
+        provider.client.chat = AsyncMock(return_value=empty_async_iter())
+
+        result = await provider._acompletion(
+            CompletionParams(
+                model_id="llama3.1",
+                messages=[{"role": "user", "content": "Hello"}],
+                response_format={"type": "json_object"},
+                stream=True,
+            ),
+        )
+        async for _ in result:  # type: ignore[union-attr]
+            pass
+
+        call_kwargs = provider.client.chat.call_args[1]
+        assert call_kwargs["format"] == "json"
 
 
 @pytest.mark.asyncio

--- a/tests/unit/providers/test_ollama_provider.py
+++ b/tests/unit/providers/test_ollama_provider.py
@@ -91,6 +91,41 @@ async def test_completion_with_dict_response_format() -> None:
 
 
 @pytest.mark.asyncio
+async def test_streaming_completion_passes_format_top_level() -> None:
+    """Test that response_format is converted and passed as a top-level `format` kwarg when streaming."""
+    from dataclasses import dataclass
+
+    @dataclass
+    class TestOutput:
+        name: str
+
+    async def empty_async_iter() -> AsyncIterator[None]:
+        return
+        yield
+
+    with patch.object(OllamaProvider, "_init_client"):
+        provider = OllamaProvider(api_key=None)
+        provider.client = Mock()
+        provider.client.chat = AsyncMock(return_value=empty_async_iter())
+
+        result = await provider._acompletion(
+            CompletionParams(
+                model_id="llama3.1",
+                messages=[{"role": "user", "content": "Hello"}],
+                response_format=TestOutput,
+                stream=True,
+            ),
+        )
+        async for _ in result:  # type: ignore[union-attr]
+            pass
+
+        call_kwargs = provider.client.chat.call_args[1]
+        assert isinstance(call_kwargs["format"], dict)
+        assert "name" in call_kwargs["format"]["properties"]
+        assert "format" not in call_kwargs.get("options", {})
+
+
+@pytest.mark.asyncio
 async def test_streaming_completion_passes_tools_top_level() -> None:
     """Test that tools are passed as a top-level kwarg to client.chat(), not inside options."""
     tools = [

